### PR TITLE
User internal event

### PR DESCRIPTION
**Pull request by mdpye**

### CODE REVIEW

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt [hunk 1]
```diff
@@ -1,164 +1,118 @@
...
-                    // That is the relevant event from the perspective of this store.
-                    // The ChatEvent.NewReadCursor will be translated
-                    // into if CursorStore updates applying ReadStateUpdatedEvent and
-                    // returning it with the cursor.
-                    val eventWithNoCursor = event.copy(event.readState.copy(cursor = null))
-                    listOf(eventWithNoCursor)
+                    this.readStates[event.readState.roomId] = event.readState
+                    listOf(
+                            UserInternalEvent.RoomUpdated(this[event.readState.roomId]!!),
+                            UserInternalEvent.NewCursor(event.readState.cursor!!) // can this be null?
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt]: I think it can. A cursor can be not set (yet) for the current user, but new messages are being sent by other users (unread count increments). We should only return `NewCursor` event if there was one that came with the event.

**Comment by mdpye** [comment]
There is still an outstanding issue with this PR: the removal of immediate application of room created, joined, left and deleted when the synchronous call completes.

This is made more tricky because of the inconsistent API types returned from the backend. :(

**Comment by mdpye** [comment]
> There is still an outstanding issue with this PR: the removal of immediate application of room created, joined, left and deleted when the synchronous call completes.

This should be resolved by https://github.com/pusher/chatkit-android/pull/144/commits/3792824c9fd9b1f9439f6079155462814a5c1a0d, but tests won't pass until the backend is deployed with v7 changes currently in review.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/SynchronousChatManager.kt [hunk 2]
```diff
@@ -221,45 +223,35 @@ class SynchronousChatManager : AppHookListener {
...
-                is UserSubscriptionEvent.RoomDeletedEvent ->
+                is UserInternalEvent.RoomDeleted->
                     ChatEvent.RoomDeleted(event.roomId)
-                is UserSubscriptionEvent.ReadStateUpdatedEvent -> {
-                    val receivedCursor = event.readState.cursor
-                    if (receivedCursor != null) {
-                        ChatEvent.NewReadCursor(receivedCursor)
-                    } else {
-                        ChatEvent.RoomUpdated(roomService.roomStore[event.readState.roomId]!!)
-                    }
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/SynchronousChatManager.kt]: Awesome to see that simplified :)

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/SynchronousCurrentUser.kt [hunk 3]
```diff
@@ -28,7 +28,7 @@ class SynchronousCurrentUser(
         private val pushNotifications: PushNotifications?
 ) {
 
-    val rooms: List<Room> get() = chatManager.roomService.roomStore.toList()
+    val rooms: List<Room> get() = chatManager.roomService.roomStore.listAll()
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/SynchronousCurrentUser.kt]: What's the motivation for this change? Kotlin convention is to call it `toList`: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/to-list.html

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        readState: ReadStateApiType?
+) = mapToRoom(
+        room = room,
+        memberships = memberships?.userIds.orEmpty().toSet(),
+        unreadCount = readState?.unreadCount
+)
+
+internal fun mapToRoom(
+        room: RoomApiType,
+        memberships: Set<String>?,
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: `memberships` should be always there for this method call

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        createdById = room.createdById,
+        pushNotificationTitleOverride = room.pushNotificationTitleOverride,
+        isPrivate = room.private,
+        customData = room.customData,
+        createdAt = room.createdAt,
+        updatedAt = room.updatedAt,
+        deletedAt = room.deletedAt,
+        lastMessageAt = room.lastMessageAt,
+        unreadCount = unreadCount,
+        memberUserIds = memberships.orEmpty()
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: using named params just introduces noise when it's clear what each param is in a given context (at a given call site)

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import com.pusher.chatkit.model.network.RoomMembershipApiType
+import com.pusher.chatkit.rooms.Room
+
+internal fun mapToRoom(
+        room: RoomApiType,
+        memberships: RoomMembershipApiType?,
+        readState: ReadStateApiType?
+) = mapToRoom(
+        room = room,
+        memberships = memberships?.userIds.orEmpty().toSet(),
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: `memberships` should not be optional and there should not be `orEmpty` call as we are using this mapping only for joined rooms hence this information should be always present

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package com.pusher.chatkit.model.mappers
+
+import com.pusher.chatkit.model.network.ReadStateApiType
+import com.pusher.chatkit.model.network.RoomApiType
+import com.pusher.chatkit.model.network.RoomMembershipApiType
+import com.pusher.chatkit.rooms.Room
+
+internal fun mapToRoom(
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: Better packaging: `com.pusher.chatkit.rooms.api.JoinedRoomApiMapper`. A class so that it can be created once (in the future by DI) and more easily swapped if needed in the future. Also having a class conveys more needed information about what are mapping from and makes it easier / more natural to navigate to (ctrl+n).

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/SynchronousCurrentUser.kt (hunk 3 above)
Comment (mdpye) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/SynchronousCurrentUser.kt]: I don't remember any more. I'll revert.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt (hunk 4 above)
Comment (mdpye) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: Heh, interesting one. The server api `get_rooms` call doesn't return memberships, but isn't implemented in the SDK (which is probably why I made it optional - knowing that not all endpoints return them).

It can probably be required if we aren't going to use that endpoint from the device.

Comment (mdpye) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: It does raise the question of how we should report the potential invalid state that we have a room entity, but not matching membership entity when we go to map it.

Comment (marekoid) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: That would mean our programming mistake and we should let the runtime to report it in the standard way such mistakes are reported, so that we see it bold and clear as soon as it's introduced and fix it immediately (having it reported close to the root cause).

If there is a chance that optionality for members of joined rooms may need to come back, then it can be done in one place, that is on the backend by the backend returning us an empty list. It's better than putting code with bug-hiding potential in all the SDKs. Also on the backend changes can be done/undone quickly at our discretion which is not the case for SDKs (where users need to upgrade).

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        room: RoomApiType,
+        memberships: RoomMembershipApiType?,
+        readState: ReadStateApiType?
+) = mapToRoom(
+        room = room,
+        memberships = memberships?.userIds.orEmpty().toSet(),
+        unreadCount = readState?.unreadCount
+)
+
+internal fun mapToRoom(
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: [Likewise](https://github.com/pusher/chatkit-android/pull/144/files#r366860798), better packaging: `com.pusher.chatkit.rooms.internal.JoinedRoomInternalMapper`.

There should be also `NonJoinedRoomApilMapper` for which we would explicitly see that `membership` and `read_status` are not there.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/network/Room.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ * SYNC REQUESTS
+ */
+
+internal data class UpdateRoomRequest(
+        val name: String?,
+        val private: Boolean?,
+        val customData: CustomData?
+)
+
+internal data class UpdateRoomRequestWithPNTitleOverride(
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/model/network/Room.kt]: Couldn't we have one class and Gson set up to not serialise nulls? I guess the motivation for this repeated class with such clumsy name is that the backend is not happy receiving null, is that correct?

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/network/Room.kt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        val pushNotificationTitleOverride: String?,
+        val private: Boolean,
+        val createdById: String,
+        val customData: CustomData?,
+        var userIds: List<String> = emptyList()
+)
+
+/*
+ * SYNC RESPONSES
+ */
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/model/network/Room.kt]: A separate `CreateRoom.kt` file in `rooms.api` package would be better (grouping two classes req-res aids in legibility) than such comment.

Also "SYNC" rather adds confusion (syncing? synchronous? -- whilst indeed they're async) than useful information. It is clear what a request/response is in HTTP (it's the subscription which is a custom thing).

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/Room.kt [hunk 11]
```diff
@@ -17,22 +17,4 @@ data class Room(
         var updatedAt: String,
         var deletedAt: String?,
         val memberUserIds: Set<String>
-) {
-
-    override fun equals(other: Any?) = (other is Room) && id == other.id
-
-    override fun hashCode(): Int { return id.hashCode() }
-
-    fun deepEquals(other: Room) =
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/Room.kt]: So glad to see that gone :)

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomService.kt [hunk 12]
```diff
@@ -78,33 +88,33 @@ internal class RoomService(
...
-                    .flatMap { body -> client.doPost<Room>("/rooms", body) }
-                    .map { room ->
-                        roomStore += room
-                        userService.populateUserStore(room.memberUserIds)
-                        room
+                    .flatMap { body -> client.doPost<CreateRoomResponse>("/rooms", body) }
+                    .map { response ->
+                        roomStore.add(response)
+                        userService.populateUserStore(response.membership.userIds.toSet())
+                        mapToRoom(response.room, response.membership, null)
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomService.kt]: When creating a room shouldn't unread count be known and equal to 0?

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomService.kt [hunk 13]
```diff
@@ -78,33 +88,33 @@ internal class RoomService(
...
-            client.doPost<Room>("/users/${URLEncoder.encode(userId, "UTF-8")}/rooms/${URLEncoder.encode(roomId, "UTF-8")}/join")
-                    .map { room ->
-                        roomStore += room
-                        userService.populateUserStore(room.memberUserIds)
-                        room
+            client.doPost<JoinRoomResponse>("/users/${URLEncoder.encode(userId, "UTF-8")}/rooms/${URLEncoder.encode(roomId, "UTF-8")}/join")
+                    .map { response ->
+                        roomStore.add(response)
+                        userService.populateUserStore(response.membership.userIds.toSet())
+                        mapToRoom(response.room, response.membership, null)
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomService.kt]: When we join a room shouldn't we get unread count?

Comment (mdpye) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomService.kt]: You can argue that we should, but IRL, we do not.
The unread count will follow immediately after in the subscription addedToRoom event, and we don't want to calculate it twice at considerable cost on the backend.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomService.kt (hunk 12 above)
Comment (mdpye) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomService.kt]: You could infer that from the fact that the room has just been created, yes, but it's not part of the backend response.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt (hunk 5 above)
Comment (mdpye) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: I find it safer/easier to refactor, especially when there is a run of `String`s for example (the various dates), but I don't feel particularly strongly about it.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt (hunk 1 above)
Comment (mdpye) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt]: This has been resolved (the room store should have nothing to do with cursors, and it no longer does)

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt [hunk 14]
```diff
@@ -1,163 +1,155 @@
...
     }
 
-    fun initialiseContents(rooms: List<Room>) {
-        clear()
-        rooms.forEach { this += it }
+    internal fun remove(roomId: String) {
+        synchronized(this) {
+            this.rooms.remove(roomId)
+            this.members.remove(roomId)
+            this.unreadCounts.remove(roomId)
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt]: no need for all those `this.` prefixes

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt [hunk 15]
```diff
@@ -1,163 +1,155 @@
...
-                    val room = roomsMap[event.roomId]!!
-                    this += room.withAddedMember(event.userId)
-                    listOf(event)
+
+                addedEvents + removedEvents + updatedEvents + membershipEvents
+            }
+            is UserSubscriptionEvent.AddedToRoomEvent -> {
+                this.rooms[event.room.id] = event.room
+                this.members[event.membership.roomId] = event.membership.userIds.toSet()
+                this.unreadCounts[event.readState.roomId] = event.readState.unreadCount
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt]: All the `this.` prefixes are not needed. I don't think they improve signal-to-noise ratio.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt [hunk 16]
```diff
@@ -1,163 +1,155 @@
...
+            }
+            is UserSubscriptionEvent.RemovedFromRoomEvent -> {
+                this.remove(event.roomId)
+                listOf(UserInternalEvent.RemovedFromRoom(event.roomId))
+            }
+            is UserSubscriptionEvent.UserJoinedRoomEvent -> {
+                this.members[event.roomId]?.let { existingMembers ->
+                    this.members[event.roomId] =
+                            existingMembers + event.userId
                 }
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt]: Room has to be there:
```kotlin
                members[event.roomId] = members[event.roomId]!! + event.userId
```

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/users/UserSubscriptionEvent.kt [hunk 17]
```diff
@@ -11,93 +13,47 @@ internal sealed class UserSubscriptionEvent {
...
 
-internal data class ReadStateApiType(
-        val roomId: String,
-        val unreadCount: Int,
-        val cursor: Cursor?
-)
-
-internal data class RoomMembershipApiType(val roomId: String, val userIds: List<String>)
\ No newline at end of file
+internal sealed class UserInternalEvent {
```
Comment (marekoid) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/users/UserSubscriptionEvent.kt]: I would put it in a separate file.

File: chatkit-core/src/test/kotlin/com/pusher/chatkit/users/UserSubscriptionEventParserSpec.kt [hunk 18]
```diff
@@ -25,36 +25,16 @@ object UserSubscriptionEventParserSpec : Spek({
         }
 
     }
-    describe("when parsing initial state with read state missing for the second room " +
-            "(simulating the top 1000 rooms limit)") {
```
Comment (marekoid) [inline chatkit-core/src/test/kotlin/com/pusher/chatkit/users/UserSubscriptionEventParserSpec.kt]: why is this test gone?

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt (hunk 5 above)
Comment (marekoid) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/model/mappers/Room.kt]: The IDE has "Change Signature" refactor option and will do it for you.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/model/network/Room.kt (hunk 9 above)
Comment (marekoid) [reply chatkit-core/src/main/kotlin/com/pusher/chatkit/model/network/Room.kt]: That's existing logic (now those classes just get a better home) and it seems `null` has to be sent so no much value in looking for any improvement here now.

**Comment by marekoid** [comment]
Thanks Mike for your good suggestions, encouragement and support in refactoring (and the nice discussion/brainstorming around it)!
